src/dice.class.php

```diff
@@ -46,22 +46,27 @@ function connectDatabase() {
 	 */
 	function requireMailsAreRegistered(array $emails) {
 		$this->connectDatabase();
-		$emails_string = "'" . implode("', '", $emails) . "'";
-		$sql = "SELECT registered_email FROM dice_emails WHERE registered_email IN ($emails_string)";
-		$result = $this->dbconn->query($sql) or exit("fatal error: data connection lost @requireMailsAreRegistered!");
-		$registered_mails = $result->fetch_array();
-		if ($result->num_rows === count($emails)) {
-			return;	// all emails are registered
+		$emails = array_unique($emails);
+		$emailCount = count($emails);
+		$placeholder = implode(", ", array_fill(0, $emailCount, "?"));
+		$sql = "SELECT COUNT(*), registered_email FROM dice_emails WHERE registered_email IN ($placeholder) GROUP BY registered_email";
```

Comment:
I'm not sure this works as intended (but the code below still produces correct behavior).  Assuming all emails are registered, won't the `COUNT(*)` returned in each record be `1` (due to the `GROUP BY`) rather than the total number of records returned?

If so, that simply means the short-circuit test in the while loop below will never be true, and all records will be stored in `result`.  I actually think the old code was clearer in this respect when it used `fetch_array()`.

Reply:
Actually, I misread the code.  Without the short circuit, you end up hitting the final `exit()` when all emails are registered.

Reply:
While debugging #33, I came up with an alternate query that offloads the logic for finding the missing emails to the DB.  Just offering it up here in case you're interested.  (Note that this still does not solve #33 due to another issue, which I'll explain over there.)

```php
	function requireMailsAreRegistered(array $emails) {
		$this->connectDatabase();
		$emails = array_unique($emails);
		$emailCount = count($emails);
		$sql = "SELECT email FROM ("
				. implode(" UNION ALL ", array_fill(0, $emailCount, "SELECT ? email"))
				. ") emails WHERE email NOT IN (SELECT registered_email FROM dice_emails)";
		$statement = $this->dbconn->prepare($sql);
		$statement->bind_param(str_repeat("s", $emailCount), ...$emails);
		$statement->execute() or exit("fatal error: data connection lost @requireMailsAreRegistered!");
		$statement->bind_result($missingEmail);
		$missingEmails = [];
		while ($statement->fetch()) {
			$missingEmails[] = $missingEmail;
		}
		if (!empty($missingEmails)) {
			exit("fatal error: emails " . implode(", ", $missingEmails) . " are not registered. Please register those emails at {$this->domain}/register.php !");
		}
	}
```

Or as a diff:

```diff
diff --git a/src/dice.class.php b/src/dice.class.php
index 69923a7..301f917 100644
--- a/src/dice.class.php
+++ b/src/dice.class.php
@@ -48,27 +48,20 @@ class dice {
                $this->connectDatabase();
                $emails = array_unique($emails);
                $emailCount = count($emails);
-               $placeholder = implode(", ", array_fill(0, $emailCount, "?"));
-               $sql = "SELECT COUNT(*), registered_email FROM dice_emails WHERE registered_email IN ($placeholder) GROUP BY registered_email";
+               $sql = "SELECT email FROM ("
+                               . implode(" UNION ALL ", array_fill(0, $emailCount, "SELECT ? email"))
+                               . ") emails WHERE email NOT IN (SELECT registered_email FROM dice_emails)";
                $statement = $this->dbconn->prepare($sql);
                $statement->bind_param(str_repeat("s", $emailCount), ...$emails);
                $statement->execute() or exit("fatal error: data connection lost @requireMailsAreRegistered!");
-               $statement->bind_result($count, $email);
-               $result = [];
+               $statement->bind_result($missingEmail);
+               $missingEmails = [];
                while ($statement->fetch()) {
-                       if ($count === $emailCount) {
-                               return; // all emails are registered
-                       }
-                       $result[] = $email;
-               }
-               if ($statement->num_rows === 0) {
-                       exit("fatal error: none of the emails is registered. Please register emails at {$this->domain}/register.php !");
+                       $missingEmails[] = $missingEmail;
                }
-               $missingEmails = array_diff($emails, $result);
                if (!empty($missingEmails)) {
                        exit("fatal error: emails " . implode(", ", $missingEmails) . " are not registered. Please register those emails at {$this->domain}/register.php !");
                }
-               exit("fatal error: unknown error with email adresses!");
        }
 
        /**
```

Reply:
@ssoloff I like your query. I like offloading algorithms to SQL in general.
This reminded me of the [longest query I ever created myself](https://gist.github.com/RoiEXLab/4ce5ab716dc16a23666e4c88e0483d7a), in case you're interested ^^
It basically queries comments from a video and sorts them by relevance ^^